### ISSUE TITLE
[Feature Request] Handling of unserializable policies

### ISSUE THREAD

**Issue by TheRisenPhoenix** [pytorch/rl#2082]
## Motivation

For my policy, I'm using a model which is, for certain reasons, tied to a hardware instance and is not serializable.
This results in not being able to use the `SyncDataCollector` (or, as it seems to me on first sight, any DataCollector), because of a `deepcopy` statement. To be precise, in `line 321` of `collectors.py`, the policy is deepcopied. Deepcopy relies on pickle, which tries to serialize the model. If it isn't serializable, an exception is thrown.

Is there a way around copying the policy? In my understanding, if the model is already initialized on the target device, why does it need to be reinitialized there (but I'm not sure whether I understood what's happening in this method)

I'm also not sure if thats the only critical copying, or whether this happens in another place too.

Is there a way to handle models that cannot be serialized? Or is this a mandatory requirement?

Any advice would be greatly appreciated! :)

## Checklist

- [ x ] I have checked that there is no similar issue in the repo (**required**)

**Comment by TheRisenPhoenix** [comment]
Thats great to hear, thank you for adding it to the backlog! I'm really looking forward that.

**Comment by vmoens** [comment]
If I run
```python
from tensordict.nn import TensorDictModuleBase as Base
from torchrl.envs import GymEnv

class NonSerializablePolicy(Base):
    in_keys = ["observation"]
    out_keys = ["action"]
    def __call__(self, data):
        data.set("action", 1)
        return data
    def __deepcopy__(self, memo):
        raise RuntimeError

policy = NonSerializablePolicy()
env = GymEnv("CartPole-v1")
print(env.rollout(10, policy))

from torchrl.collectors import SyncDataCollector

c = SyncDataCollector(env, policy=policy, policy_device=None, total_frames=100, frames_per_batch=10)
for d in c:
    print(d)
```
It runs ok, but if I tell the collector that I want the policy to be on "cpu" it crashes because i'm casting it to device.

Question to you @TheRisenPhoenix : do you need a device for your policy? Presumably if it's not copyable it isn't a nn.Module, or is it?

Would that work for you that we just capture the error and tell the user to set policy_device to None?

**Comment by TheRisenPhoenix** [comment]
Hi @vmoens and thanks for the quick investigation!
It's true, giving the `SyncDataCollector`  `None` devices avoids the crash at this point, unfortunately it then fails later (for me when using `DiscreteSACLoss`, but probably this is not the only place where the policy is deepcopied... 😢). More specifically, the problem here seems to be in `convert_to_functional` (`common.py:302`). Do you have any suggestions on how to get around this?

My model is actually a `nn.Module`, but uses a custom wrapper layer inside to connect to the hardware I'm using, so it's not completely torchy.

**Comment by vmoens** [comment]
Ah got it, interesting
So we defo need to fix this then!
For context, we scrape out the params from the model and make a stateless copy of the model that we then use to make functional calls without perturbing your original model.
I guess we could consider two options:
- keep the functional calls but no copy of the model - I think this could be dangerous if you modify your model outside of the loss (e.g. during data collection)
- disable functional calls for these models (same drawback as above)
It's something we can do on a per-model basis.

By the way: Is your value model copyable?

**Comment by TheRisenPhoenix** [comment]
Ah good point, my value model is copyable at the moment, but that was just for testing purposes, so it would be great if the same fixes applied to that too!

What could be a valid reason to modify the model outside of the loss?

**Comment by vmoens** [comment]
This should solve the problem for losses too
https://github.com/pytorch/rl/pull/2148

**Comment by TheRisenPhoenix** [comment]
Awesome, thanks a lot! 🥳
